File: algorithm/backtracking/python/letter_combination_phone_number.py
"""
Letter Combinations of a Phone Number
https://leetcode.com/problems/letter-combinations-of-a-phone-number

Given a string containing digits from 2-9 inclusive, return all possible letter
combinations that the number could represent.

Approaches:
    1. letter_combination (DFS / backtracking)
       Time: O(4^n), Space: O(n) call stack (result excluded).
    2. letter_combination_bfs (BFS / iterative build-up)
       Time: O(4^n), Space: O(4^n).
"""

from typing import List


mapping = {
    '2': "abc",
    '3': "def",
    '4': "ghi",
    '5': "jkl",
    '6': "mno",
    '7': "pqrs",
    '8': "tuv",
    '9': "wxyz",
}


def _dfs(idx: int, digits: str, res: List[str], curr: List[str]) -> None:
    if idx == len(digits):
        res.append("".join(curr))
        return

    for c in mapping[digits[idx]]:
        curr.append(c)
        _dfs(idx + 1, digits, res, curr)
        curr.pop()


# Method 1: DFS
# Time: O(4^n), Space: O(n), where n = len(digits)
# We have up to 4 letters for each digit, and we have n digits.
# Space: O(n) because of the call stack.
# The result list space if counted, is O(4^n).
def letter_combination(digits: str) -> List[str]:
    if not digits:
        return []

    res: List[str] = []
    _dfs(0, digits, res, [])
    return res


# Method 2: BFS
# Time: O(4^n), Space: O(4^n)
def letter_combination_bfs(digits: str) -> List[str]:
    if not digits:
        return []

    res: List[str] = [""]

    for digit in digits:
        temp: List[str] = []
        for s in res:
            for c in mapping[digit]:
                temp.append(s + c)
        res = temp

    return res


if __name__ == "__main__":
    res = letter_combination("23")
    expected = ["ad", "ae", "af", "bd", "be", "bf", "cd", "ce", "cf"]
    assert res == expected
    res = letter_combination_bfs("23")
    assert res == expected

    res = letter_combination("")
    expected = []
    assert res == expected
    res = letter_combination_bfs("")
    assert res == expected

    res = letter_combination("2")
    expected = ["a", "b", "c"]
    assert res == expected
    res = letter_combination_bfs("2")
    assert res == expected

File: algorithm/backtracking/cpp/letter_combination_phone_number.cpp
// https://leetcode.com/problems/letter-combinations-of-a-phone-number

/* 
Given a string containing digits from 2-9 inclusive, return all possible letter combinations that the number could represent. Return the answer in any order.
A mapping of digits to letters (just like on the telephone buttons) is given below. Note that 1 does not map to any letters.

Ex1:
Input: digits = "23"
Output: ["ad","ae","af","bd","be","bf","cd","ce","cf"]

Ex2:
Input: digits = ""
Output: []


Ex3:
Input: digits = "2"
Output: ["a","b","c"]

*/

#include <vector>
#include <unordered_map>
#include <iostream>
#include <cassert>

using namespace std;

unordered_map<char, string> mapping {{'2', "abc"},
                                    {'3', "def"},
                                    {'4', "ghi"},
                                    {'5', "jkl"},
                                    {'6', "mno"},
                                    {'7', "pqrs"},
                                    {'8', "tuv"},
                                    {'9', "wxyz"}};

void dfs(int idx, string& digits, vector<string>& res, string curr) {
    if (idx == digits.size()) {
        res.push_back(curr);
        return;
    }

    for (char c: mapping[digits[idx]]) {
        curr.push_back(c);
        dfs(idx+1, digits, res, curr);
        curr.pop_back();
    }
}

// Method 1: DFS
// Time: O(4^n), Space: O(n), where n = digits.size()
// We have 4 letters for each digit, and we have n digits.
// Space: O(n) because of the call stack.
// The result vector space if counted, is O(4^n).
vector<string> letterCombination(string digits) {
    if (digits.empty()) {
        return {};
    }

    vector<string> res;
    dfs(0, digits, res, "");
    return res;
}

// Method 2: BFS
// Time: O(4^n), Space: O(4^n)
vector<string> letterCombinationBFS(string digits) {
    if (digits.empty()) {
        return {};
    }

    vector<string> res;
    res.push_back("");

    for (char digit: digits) {
        vector<string> temp;
        for (string s: res) {
            for (char c: mapping[digit]) {
                temp.push_back(s+c);
            }
        }
        res = temp;
    }

    return res;
}

int main() {
    vector<string> res = letterCombination("23");
    vector<string> expected = {"ad","ae","af","bd","be","bf","cd","ce","cf"};
    assert(res == expected);
    res = letterCombinationBFS("23");
    assert(res == expected);

    res = letterCombination("");
    expected = {};
    assert(res == expected);
    res = letterCombinationBFS("");
    assert(res == expected);

    res = letterCombination("2");
    expected = {"a","b","c"};
    assert(res == expected);
    res = letterCombinationBFS("2");
    assert(res == expected);
}
